refactor: improve product barcode management

ellpanda23 opened the pull request:
## 📝 Description

This PR refactors the product entity to support multiple, active/inactive, and primary/secondary barcodes, moving away from the single `sku` field for primary identification. This change provides greater flexibility for inventory management and point-of-sale systems.

## 🚀 Changes Implemented

-   **Model:** Implemented new methods in the `Product` and `Barcode` models to manage the relationship and search logic:
    -   `Product::barcodes()`
    -   `Product::primaryBarcode()`
    -   `Barcode::findProductByCode()` (updated logic)
    -   `Barcode::getPrimaryCodeForProduct()`
    -   `Barcode::codeExists()`
-   **Service/Logic:** Updated product creation and editing workflows to correctly handle the persistence of multiple barcodes, ensuring only one is designated as 'primary' and that all are validated for uniqueness.
-   **Views/UI:** Updated the product forms and relevant views to incorporate the new structure for managing the barcode list.
-   **Translations:** Added new language keys for all updated fields and UI elements related to the new barcode structure.

- Related issue: #320 

## 🔎 How to Test

1.  Navigate to the **Product Creation** page.
2.  Attempt to create a product with multiple barcode entries, marking one as 'primary'.
3.  Verify that only one barcode can be set as primary and that all codes are unique across the database.
4.  Save the product and confirm that all barcode entries are saved correctly in the database.
5.  Use the `Barcode::findProductByCode('your_code')` logic (e.g., in a test route or console) to confirm a product can be retrieved by any of its active barcodes, or by its SKU if the barcode is not found.
6.  Edit an existing product and confirm you can add, remove, and change the primary barcode.

## ✅ Checklist

* [x] My code follows the project's code style guidelines.
* [x] I have tested my changes thoroughly (unit tests, manual testing, etc.).
* [ ] I have updated the documentation (if necessary).
* [x] All new and existing tests pass.
* [x] This change does not introduce any security vulnerabilities.
* [x] The commit message follows the Conventional Commits specification.

- Implemented a new method in the Product model to search for products by barcode or SKU.
- Updated the product creation and editing logic to handle multiple barcodes.
- Made changes to the views and forms to reflect the new barcode structure.
- Added translations for the new barcode-related fields.

## 🧾 Screenshots / Logs (if UI or relevant)

<img width="623" height="370" alt="image" src="https://github.com/user-attachments/assets/29dc6902-2a55-46ad-b10b-11b3c2d002b8" />

<img width="1250" height="886" alt="image" src="https://github.com/user-attachments/assets/8aa9ab37-16c9-42d8-abad-3f234d9f059c" />


<img width="1298" height="1021" alt="image" src="https://github.com/user-attachments/assets/ba80b118-c480-412b-9907-202ecaf93227" />

---

ellpanda23 committed:
feat(products): add barcode scanner to product form

Integrates the html5-qrcode library to introduce a barcode scanning feature in the product creation and editing forms.

Key changes include:
- Added a new `generateBarcodeFormComponent` to create the barcode input field with a camera icon action.
- The action opens a modal containing the camera view for scanning.
- On successful scan, the barcode input is automatically populated, and the modal closes.
- The scanner is correctly stopped when the modal is closed, either automatically after a scan or manually by the user, preventing the camera from remaining active."

ellpanda23 committed:
refactor(barcode-scanner): translate messages and comments to EnglishThe labels and messages in the barcode scanning component are updated to be in English. This includes changes to button texts, console messages, and comments in the code, improving language consistency in the application.

ellpanda23 committed:
refactor(cashier): improve code structure and readability

Changes were made to the cashier view to optimize code readability and structure. Styles were adjusted and the organization of elements was improved, including the implementation of a QR code scanner in a modal. Additionally, some minor bugs were fixed and event handling for barcode scanning was improved.

ellpanda23 committed:
refactor(cashier): translate comments and messages to English

Comments and messages in the cashier view component are updated to be in English, improving language consistency across the application. Adjustments were made to the code structure and the state management of the QR scanner was optimized.

ellpanda23 committed:
bug fixed Can't Save "About" in General Settings lakasir#334 and changes the logic for save images now it doesn't desapear if you only change other shop infomation

ellpanda23 committed:
Merge pull request #2 from ellpanda23/feat/barcode-scanner

Feat/barcode scanner

ellpanda23 committed:
Merge pull request #3 from ellpanda23/fix/cant-save-about-in-general-setting-tab

bug fixed Can't Save "About" in General Settings lakasir#334 and chan…

ellpanda23 committed:
feat(settings): added and option for  MXN currency in general settings

ellpanda23 committed:
Merge pull request #4 from ellpanda23/added-mxn-currency-support

feat(settings): added and option for  MXN currency in general settings

ellpanda23 committed:
refactor: improve product barcode management

- Implemented a new method in the Product model to search for products by barcode or SKU.
- Updated the product creation and editing logic to handle multiple barcodes.
- Made changes to the views and forms to reflect the new barcode structure.
- Added translations for the new barcode-related fields.

sheenazien8 commented:
Thanks for your work @ellpanda23 , this will improve Lakasir a lot! 🙌
I encountered an issue: when editing a product, I get the error "barcode already exists". My expectation is that during edit, the validation should ignore the barcode that already belongs to the current product.

<img width="787" height="982" alt="image" src="https://github.com/user-attachments/assets/5f674af7-7cab-485e-9342-2631e1b5fc0a" />

sheenazien8 commented:
and the ui i think better the name and the barcodes have fullspan like this

https://github.com/lakasir/lakasir/blob/b39171258cb1a2e11214cfe23c58ba577b06fb54/app/Filament/Tenant/Resources/ProductResource.php#L173
```php
  $this->generateNameFormComponent(),
  $this->generateBarcodeFormComponent()
      ->columnSpanFull(),
```

ellpanda23 committed:
refactor: improve barcode management in products

Adjusted the validation logic to allow excluding the current barcode ID when checking for uniqueness.

Added properties and made adjustments to load barcode data during product editing.

Improved the form structure for better presentation and functionality of barcodes.

ellpanda23 commented:
Hello @sheenazien8 I just fixed the update problem now works well, you can update the code without any issue I also added the 
`->columnSpanFull()` to the name and barcode fields.

<img width="1238" height="1013" alt="image" src="https://github.com/user-attachments/assets/fd19e9ed-3863-4e71-b81d-399ad20ae42a" />

ellpanda23 committed:
fix: correct typo in the 'secondary' option of the product form

sheenazien8 commented:
> Hello @sheenazien8 I just fixed the update problem now works well, you can update the code without any issue I also added the `->columnSpanFull()` to the name and barcode fields.
> 
> <img alt="image" width="1238" height="1013" src="https://private-user-images.githubusercontent.com/75634056/497405185-fd19e9ed-3863-4e71-b81d-399ad20ae42a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk1NzA2MzcsIm5iZiI6MTc1OTU3MDMzNywicGF0aCI6Ii83NTYzNDA1Ni80OTc0MDUxODUtZmQxOWU5ZWQtMzg2My00ZTcxLWI4MWQtMzk5YWQyMGFlNDJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA0VDA5MzIxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0YWRhMGY3Y2YxMTdjNTAzMWU3ZjQ3MWU5ZjNhMjRmYzE3M2ZkZTZmNTQ1YTlhMDZjMjJlYmMwODAxYTliOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iNDPUW_v9ZqOZbY-_9PHBTt6Jb-Da5YVD_KbNICCoz0">

thanks, i'll test it again

sheenazien8 commented:
sweet the save and find product use barcode in cashier page works properly, but i need small change again in PrintLabel, in print label i think wee need the option what's the barcode that i want to print, you can take a look in here https://github.com/lakasir/lakasir/blob/5cbe8e609c8e168c15d5492de032d5ee27f49b70/app/Filament/Tenant/Resources/ProductResource/Pages/PrintLabel.php#L21

@ellpanda23

ellpanda23 committed:
Refactor: Improve barcode handling on label printing page

Updated logic to fetch the barcode, enabling the use of a provided value instead of the primary code. Also added a selection field to choose the available barcode, enhancing form usability.

ellpanda23 commented:
I added the suggestion now you can select what barcode you want to print on the print label page, if you have any other suggestions or an issue I'll be happy to work on it.

The field is required to avoid an empty barcode
<img width="410" height="507" alt="image" src="https://github.com/user-attachments/assets/cb49d397-3b89-4751-910a-4b095eeba0df" />

<img width="1253" height="946" alt="image" src="https://github.com/user-attachments/assets/101c9fe6-6ce4-42b7-a0a5-bfbc563375fb" />
<img width="1257" height="651" alt="image" src="https://github.com/user-attachments/assets/fa940f6e-c6cf-43ea-8629-0df9fc2ab50e" />

@sheenazien8

ellpanda23 committed:
Merge branch 'local-testing' into improve/multiple-barcode

sheenazien8 commented:
sweet, that's cool, i'll merge it, thanks for your work!
